File: app/src/main/java/io/github/marcocipriani01/telescopetouch/control/RealMagneticDeclinationCalculator.java
/*
 * Copyright 2021 Marco Cipriani (@marcocipriani01)
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package io.github.marcocipriani01.telescopetouch.control;

import android.hardware.GeomagneticField;
import android.location.Location;

import androidx.annotation.NonNull;

/**
 * Encapsulates the calculation of magnetic declination for the user's location
 * and position.
 *
 * @author John Taylor
 */
public class RealMagneticDeclinationCalculator implements MagneticDeclinationCalculator {

    private GeomagneticField geomagneticField;

    /**
     * {@inheritDoc}
     * Silently returns zero if the time and location have not been set.
     */
    @Override
    public float getDeclination() {
        if (geomagneticField == null) return 0.0f;
        return geomagneticField.getDeclination();
    }

    /**
     * Sets the user's current location and time.
     */
    @Override
    public void setLocationAndTime(Location location, long timeInMillis) {
        geomagneticField = new GeomagneticField((float) location.getLatitude(),
                (float) location.getLongitude(), (float) location.getAltitude(), timeInMillis);
    }

    @NonNull
    @Override
    public String toString() {
        return "Real Magnetic Correction";
    }
}

File: app/src/main/java/com/google/android/stardroid/control/RealMagneticDeclinationCalculator.kt
// Copyright 2009 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package com.google.android.stardroid.control

import android.hardware.GeomagneticField
import com.google.android.stardroid.math.LatLong

/**
 * Encapsulates the calculation of magnetic declination for the user's location
 * and position.
 *
 * @author John Taylor
 */
class RealMagneticDeclinationCalculator : MagneticDeclinationCalculator {
    private var geomagneticField: GeomagneticField? = null

    /**
     * {@inheritDoc}
     * Silently returns zero if the time and location have not been set.
     */
    override val declination: Float
        get() = if (geomagneticField == null) {
            0.toFloat()
        } else geomagneticField!!.declination

    /**
     * Sets the user's current location and time.
     */
    override fun setLocationAndTime(location: LatLong?, timeInMillis: Long) {
        geomagneticField = GeomagneticField(
            location!!.latitude,
            location.longitude,
            0F,
            timeInMillis
        )
    }

    override fun toString(): String {
        return "Real Magnetic Correction"
    }
}